contracts/yellow-button/lib.rs

```diff
@@ -0,0 +1,545 @@
+    #[ink(event)]
+    #[derive(Debug)]
+    pub struct ButtonPressed {
+        #[ink(topic)]
+        by: AccountId,
+        score: u32,
+        total_scores: u32,
+        when: u32,
+        previous_press: u32,
+        new_deadline: u32,
+    }
+
+    /// Event emitted when TheButton owner is changed
+    #[ink(event)]
+    #[derive(Debug)]
+    pub struct OwnershipTransferred {
+        #[ink(topic)]
+        from: AccountId,
+        #[ink(topic)]
+        to: AccountId,
+    }
+
+    /// Event emitted when TheButton is created
+    #[ink(event)]
+    #[derive(Debug)]
+    pub struct ButtonCreated {
+        #[ink(topic)]
+        button_token: AccountId,
+        start: u32,
+        deadline: u32,
+    }
+
+    /// Event emitted when account is whitelisted to play the game
+    #[ink(event)]
+    #[derive(Debug)]
+    pub struct AccountWhitelisted {
+        #[ink(topic)]
+        player: AccountId,
+    }
+
+    /// Even emitted when button death is triggered    
+    #[ink(event)]
+    #[derive(Debug)]
+    pub struct ButtonDeath {
+        #[ink(topic)]
+        pressiah: Option<AccountId>,
+        pressiah_reward: u128,
+        rewards: Vec<(AccountId, u128)>,
+    }
+
+    impl YellowButton {
+        /// Returns the buttons status
+        #[ink(message)]
+        pub fn is_dead(&self) -> bool {
+            self.is_dead
+        }
+
+        /// Returns the current deadline
+        #[ink(message)]
+        pub fn deadline(&self) -> u32 {
+            self.deadline
+        }
+
+        /// Returns the user score
+        #[ink(message)]
+        pub fn score_of(&self, user: AccountId) -> u32 {
+            self.presses.get(&user).unwrap_or(0)
+        }
+
+        /// Returns whether given account can play
+        #[ink(message)]
+        pub fn can_play(&self, user: AccountId) -> bool {
+            self.can_play.get(&user).unwrap_or(false)
+        }
+
+        /// Returns the account id that pressed as last
+        #[ink(message)]
+        pub fn last_presser(&self) -> Option<AccountId> {
+            self.last_presser
+        }
+
+        /// Returns address of the game's token
+        #[ink(message)]
+        pub fn get_button_token(&self) -> Result<AccountId> {
+            Ok(self.button_token)
+        }
+
+        /// Returns game token balance of the game contract
+        #[ink(message)]
+        pub fn get_balance(&self) -> Result<Balance> {
+            let this = self.env().account_id();
+            let button_token = self.button_token;
+
+            let balance = build_call::<DefaultEnvironment>()
+                .call_type(Call::new().callee(button_token))
+                .exec_input(ExecutionInput::new(Selector::new(BALANCE_OF_SELECTOR)).push_arg(this))
+                .returns::<Balance>()
+                .fire()?;
+
+            Ok(balance)
+        }
+
+        fn emit_event<EE: EmitEvent<YellowButton>>(emitter: EE, event: Event) {
+            emitter.emit_event(event);
+        }
```

Comment:
I don't understand the purpose of this. You mind explaining?

Reply:
TL;DR; https://github.com/paritytech/ink/issues/1281